File: problem-solving/_lc-375-medium-guess-number-higher-or-lower-ii/2-bottom-up.py
"""
https://leetcode.com/problems/guess-number-higher-or-lower-ii/

Complexity
----------
Time: O(n^3)
Space: O(n^2)
"""

class Solution:
    def getMoneyAmount(self, n: int) -> int:
        dp = [[0] * (n + 1) for _ in range(n + 1)]
        # This loop in run in reverse order because in the recurrence relation,
        # we are accessing future index (g + 1) of l.
        for l in range(n - 1, 0, -1):
            for h in range(l + 1, n + 1):
                dp[l][h] = math.inf
                for g in range(l, h):
                    temp = g + max(dp[l][g - 1], dp[g + 1][h])
                    dp[l][h] = min(dp[l][h], temp)
        return dp[1][n]

File: problem-solving/lc-1143-medium-longest-common-subsequence/2-bottom-up.go
/*
https://leetcode.com/problems/longest-common-subsequence/

Complexity
----------
Time: O(m*n)
Space: O(m*n)
*/

func longestCommonSubsequence(text1 string, text2 string) int {
    var dp [][]int
    m, n := len(text1), len(text2)
    for i := 0; i < m + 1; i++ {
        dp = append(dp, make([]int, n + 1))
    }
    for i := 1; i < m + 1; i++ {
        for j :=1; j < n + 1; j++ {
            if text1[i - 1] == text2[j - 1] {
                dp[i][j] = 1 + dp[i - 1][j - 1]
            } else {
                dp[i][j] = max(dp[i - 1][j], dp[i][j - 1])
            }
        }
    }
    return dp[m][n]
}

func max(x, y int) int {
    if x > y {
        return x
    }
    return y
}
